# **Import Libraries and dataset**

In [ ]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = "/content"
%cd /content
!kaggle datasets download -d ankitverma2010/ecommerce-customer-churn-analysis-and-prediction
!unzip \*.zip  && rm *.zip

/content
  0% 0.00/532k [00:00<?, ?B/s]
100% 532k/532k [00:00<00:00, 28.8MB/s]
Archive:  ecommerce-customer-churn-analysis-and-prediction.zip
  inflating: E Commerce Dataset.xlsx  


In [ ]:
import logging

logger = logging.getLogger()

file_handler = logging.FileHandler(filename='ecc_dc_log.log', mode='w')
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
file_handler.setFormatter(formatter)
logger.addHandler(file_handler)

logger.setLevel(logging.DEBUG)
logger.info('FILE STARTS RUNNING!')

INFO:root:FILE STARTS RUNNING!


In [ ]:
import pandas as pd
import numpy as np

INFO:numexpr.utils:NumExpr defaulting to 2 threads.


In [ ]:
logger.info('Libraries imported.')

INFO:root:Libraries imported.


In [ ]:
# Reading dataset to a dataframe
df = pd.read_excel('/content/E Commerce Dataset.xlsx', sheet_name='E Comm')
df.head()

CustomerID  Churn  Tenure PreferredLoginDevice  CityTier  WarehouseToHome  \
0       50001      1     4.0         Mobile Phone         3              6.0   
1       50002      1     NaN                Phone         1              8.0   
2       50003      1     NaN                Phone         1             30.0   
3       50004      1     0.0                Phone         3             15.0   
4       50005      1     0.0                Phone         1             12.0   

  PreferredPaymentMode  Gender  HourSpendOnApp  NumberOfDeviceRegistered  \
0           Debit Card  Female             3.0                         3   
1                  UPI    Male             3.0                         4   
2           Debit Card    Male             2.0                         4   
3           Debit Card    Male             2.0                         4   
4                   CC    Male             NaN                         3   

     PreferedOrderCat  SatisfactionScore MaritalStatus  NumberOfAddress  \
0  Laptop & Accessory                  2        Single                9   
1              Mobile                  3        Single                7   
2              Mobile                  3        Single                6   
3  Laptop & Accessory                  5        Single                8   
4              Mobile                  5        Single                3   

   Complain  OrderAmountHikeFromlastYear  CouponUsed  OrderCount  \
0         1                         11.0         1.0         1.0   
1         1                         15.0         0.0         1.0   
2         1                         14.0         0.0         1.0   
3         0                         23.0         0.0         1.0   
4         0                         11.0         1.0         1.0   

   DaySinceLastOrder  CashbackAmount  
0                5.0          159.93  
1                0.0          120.90  
2                3.0          120.28  
3                3.0          134.07  
4                3.0          129.60

In [ ]:
logger.info('Data loaded.')

INFO:root:Data loaded.


In [ ]:
# Reading data information/dictionary to a dataframe
data_info = pd.read_excel('/content/E Commerce Dataset.xlsx', sheet_name=0)
data_info.drop(['Unnamed: 0', 'Unnamed: 1'], axis=1, inplace=True)
data_info

Unnamed: 2  \
0                      Variable   
1                    CustomerID   
2                         Churn   
3                        Tenure   
4          PreferredLoginDevice   
5                      CityTier   
6               WarehouseToHome   
7          PreferredPaymentMode   
8                        Gender   
9                HourSpendOnApp   
10     NumberOfDeviceRegistered   
11             PreferedOrderCat   
12            SatisfactionScore   
13                MaritalStatus   
14              NumberOfAddress   
15                     Complain   
16  OrderAmountHikeFromlastYear   
17                   CouponUsed   
18                   OrderCount   
19            DaySinceLastOrder   
20               CashbackAmount   

                                           Unnamed: 3  
0                                         Discerption  
1                                  Unique customer ID  
2                                          Churn Flag  
3                  Tenure of customer in organization  
4                  Preferred login device of customer  
5                                           City tier  
6   Distance in between warehouse to home of customer  
7                Preferred payment method of customer  
8                                  Gender of customer  
9   Number of hours spend on mobile application or...  
10  Total number of deceives is registered on part...  
11  Preferred order category of customer in last m...  
12          Satisfactory score of customer on service  
13                         Marital status of customer  
14  Total number of added added on particular cust...  
15        Any complaint has been raised in last month  
16       Percentage increases in order from last year  
17  Total number of coupon has been used in last m...  
18  Total number of orders has been places in last...  
19                   Day Since last order by customer  
20                     Average cashback in last month

In [ ]:
# Saving it to a txt for easy reference
np.savetxt(r'data_info.txt', data_info.values, fmt='%s')

# **Data Cleaning**

In [ ]:
logger.info('Data cleaning starts')

INFO:root:Data cleaning starts


## **Basic Info**

In [ ]:
# Basic information on the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5630 entries, 0 to 5629
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   CustomerID                   5630 non-null   int64  
 1   Churn                        5630 non-null   int64  
 2   Tenure                       5366 non-null   float64
 3   PreferredLoginDevice         5630 non-null   object 
 4   CityTier                     5630 non-null   int64  
 5   WarehouseToHome              5379 non-null   float64
 6   PreferredPaymentMode         5630 non-null   object 
 7   Gender                       5630 non-null   object 
 8   HourSpendOnApp               5375 non-null   float64
 9   NumberOfDeviceRegistered     5630 non-null   int64  
 10  PreferedOrderCat             5630 non-null   object 
 11  SatisfactionScore            5630 non-null   int64  
 12  MaritalStatus                5630 non-null   object 
 13  NumberOfAddress   

**OBSERVATIONS**
- 5630 entries and 20 columns
- **Churn** is our target variable
- 8 floating point variables, 7 integer variables and 5 string variables
- There are some missing values.

**Check for duplicate rows** -

In [ ]:
df.duplicated().sum()

0

No duplicates

## **Check for Inconsistancies and corruptness**

In [ ]:
# Find the unique number of values for each variable
df.nunique()

CustomerID                     5630
Churn                             2
Tenure                           36
PreferredLoginDevice              3
CityTier                          3
WarehouseToHome                  34
PreferredPaymentMode              7
Gender                            2
HourSpendOnApp                    6
NumberOfDeviceRegistered          6
PreferedOrderCat                  6
SatisfactionScore                 5
MaritalStatus                     3
NumberOfAddress                  15
Complain                          2
OrderAmountHikeFromlastYear      16
CouponUsed                       17
OrderCount                       16
DaySinceLastOrder                22
CashbackAmount                 2586
dtype: int64

Except for Customer ID and CashbackAmount, lets find the unique values for the other columns

In [ ]:
# Unique values for columns
for col in df.drop(['CustomerID', 'CashbackAmount'], axis=1).columns:
  print('Column Name :', col)
  print(df[col].unique())
  print('-------------------------------------------------------------')

Column Name : Churn
[1 0]
-------------------------------------------------------------
Column Name : Tenure
[ 4. nan  0. 13. 11.  9. 19. 20. 14.  8. 18.  5.  2. 30.  1. 23.  3. 29.
  6. 26. 28.  7. 24. 25. 10. 15. 22. 27. 16. 12. 21. 17. 50. 60. 31. 51.
 61.]
-------------------------------------------------------------
Column Name : PreferredLoginDevice
['Mobile Phone' 'Phone' 'Computer']
-------------------------------------------------------------
Column Name : CityTier
[3 1 2]
-------------------------------------------------------------
Column Name : WarehouseToHome
[  6.   8.  30.  15.  12.  22.  11.   9.  31.  18.  13.  20.  29.  28.
  26.  14.  nan  10.  27.  17.  23.  33.  19.  35.  24.  16.  25.  32.
  34.   5.  21. 126.   7.  36. 127.]
-------------------------------------------------------------
Column Name : PreferredPaymentMode
['Debit Card' 'UPI' 'CC' 'Cash on Delivery' 'E wallet' 'COD' 'Credit Card']
-------------------------------------------------------------
Column 

**OBSERVATIONS :** We can observe the following inconsistancies:
- **PreferredLoginDevice** : 'Mobile Phone' and 'Mobile' probably indicates mobile phone, since you cannot login with a land phone -> change it to a common 'Mobile'
- **WarehouseToHome** : Has extremely large values like 126 and 127, which is highly unlikely since people don't usually travel that long distance to buy goods. The values might be mistakenly entered -> change 126->26 and 127->27
- **PreferredPaymentMode** : 
  - 'CC' and 'Credit Card' indicate credit card -> change it to 'CC'
  - 'Cash on Delivery' and 'COD' indicate cash on delivery -> change it to 'COD'
  - In addition, change Debit Card to DC for convenience.
- **PreferedOrderCat**:
  - 'Mobile' and 'Mobile Phone' -> change it to 'Mobile'
  - 'Laptop & Accessory' -> change it to 'Laptop' for convenience.
- **NumberOfAddress** : Has large numbers above 10, which seems very unlikely. Need to look into it.

In [ ]:
# Correct values of column PreferredLoginDevice, WarehouseToHome, PreferredPaymentMode and PreferedOrderCat
df['PreferredLoginDevice'].replace({'Mobile Phone':'Mobile', 'Phone':'Mobile'}, inplace=True)
df['WarehouseToHome'].replace({126:26, 127:27}, inplace=True)
df['PreferredPaymentMode'].replace({'Credit Card':'CC', 'Cash on Delivery':'COD', 'Debit Card':'DC' }, inplace=True)
df['PreferedOrderCat'].replace({'Mobile Phone':'Mobile', 'Laptop & Accessory':'Laptop'}, inplace=True)

Now lets delve into the problem with the inconsistant values of NumberOfAddress.

In [ ]:
# Number of rows with NumberOfAddress greater than 5
df[df.NumberOfAddress>5].shape[0]

1453

There seems to be 1453 customers with more than 5 addresses registered in the ecommerce platform, in their name. This seems very unlikely. But we cannot rule out the following possibilities:
- The customer might have multiple properties(houses) in their name.
- Slight change in the address text.
- Addresses of workplace/workplaces if the customer works shifts.
- Ordering for someone else
- Someone else must be ordering in the customer's name name(either other family member with the customer's knowledge or in the rare case fraud transactions by outsiders).

In the above mention cases, multiple addresses might be added. 

If we have some other information like the actual list of addresses registered or the locations where the orders were made, number of family members, number of children, occupation related information, annual income etc., we might get a little bit insight into the possibility of customers have multiple addresses.

However, since we do not have them, we need to explore with the available information and reach a conclusion. We shall analyse more in the EDA process and come to a conclusion.

## **Dealing with missing values**

In [ ]:
# Find columns with missing values
missing_cols = [col for col in df.columns if df[col].isnull().any()]
missing_cols

['Tenure',
 'WarehouseToHome',
 'HourSpendOnApp',
 'OrderAmountHikeFromlastYear',
 'CouponUsed',
 'OrderCount',
 'DaySinceLastOrder']

In [ ]:
# Find the % of missing value in the above list of columns
percent = df[missing_cols].isnull().sum() * 100/len(df)
missing_df = pd.DataFrame({'column':missing_cols,
                           'no. of missing values' : df[missing_cols].isnull().sum().values,
                           'percentage' : percent.values})
missing_df

column  no. of missing values  percentage
0                       Tenure                    264    4.689165
1              WarehouseToHome                    251    4.458259
2               HourSpendOnApp                    255    4.529307
3  OrderAmountHikeFromlastYear                    265    4.706927
4                   CouponUsed                    256    4.547069
5                   OrderCount                    258    4.582593
6            DaySinceLastOrder                    307    5.452931

We see that the percent of missing values for any column is below 6%. We can either drop them (columnwise/rowwise) or impute them using any suitable method or a combination of both.

Dropping the column is not recommended, since each column has a max of only 6% missing values.

Lets look at the number of missing values in each row. If any rows has 4 or more than 4 missing values, we shall drop those rows.

In [ ]:
# Calculate the number of missing values in each row and finding the count
df.isnull().sum(axis=1).value_counts()

0    3774
1    1856
dtype: int64

The maximum number of missing value in any given row is 1. So, its difficult to drop rows with losing substantial information from the data. Therefore we shall impute them.

In our case, all our missing data are numeric type. Therefore, we shall fill the missing values with mean or median of the variable, depending in the skewness value.

In [ ]:
# Calculate skewness of missing value columns
df[missing_cols].skew(axis=0, skipna=True, numeric_only=True)

Tenure                         0.736513
WarehouseToHome                0.898406
HourSpendOnApp                -0.027213
OrderAmountHikeFromlastYear    0.790785
CouponUsed                     2.545653
OrderCount                     2.196414
DaySinceLastOrder              1.191000
dtype: float64

We can see that 
- **HourSpendOnApp** is normally distributed -> Fill missing values with mean
- **Tenure and OrderAmountHikeFromlastYear** are moderately skewed -> Fill missing values with mean
- **WarehouseToHome, CouponUsed, OrderCount and DaySinceLastOrder** are highly skewed -> Fill missing values with median

In [ ]:
round(df.HourSpendOnApp.mean()), round(df.Tenure.mean()), round(df.OrderAmountHikeFromlastYear.mean()), round(df.WarehouseToHome.mean())

(3, 10, 16, 16)

In [ ]:
df.CouponUsed.median(), df.OrderCount.median(), df.DaySinceLastOrder.median()

(1.0, 2.0, 3.0)

In [ ]:
# Fill missing values ith mean and median
df.fillna({'HourSpendOnApp':round(df.HourSpendOnApp.mean()), 
           'Tenure':round(df.Tenure.mean()), 
           'OrderAmountHikeFromlastYear':round(df.OrderAmountHikeFromlastYear.mean()),
           'WarehouseToHome':round(df.WarehouseToHome.mean()), 
           'CouponUsed':df.CouponUsed.median(), 
           'OrderCount':df.OrderCount.median(), 
           'DaySinceLastOrder':df.DaySinceLastOrder.median()}, inplace=True)

In [ ]:
df.head()

CustomerID  Churn  Tenure PreferredLoginDevice  CityTier  WarehouseToHome  \
0       50001      1     4.0               Mobile         3              6.0   
1       50002      1    10.0               Mobile         1              8.0   
2       50003      1    10.0               Mobile         1             30.0   
3       50004      1     0.0               Mobile         3             15.0   
4       50005      1     0.0               Mobile         1             12.0   

  PreferredPaymentMode  Gender  HourSpendOnApp  NumberOfDeviceRegistered  \
0                   DC  Female             3.0                         3   
1                  UPI    Male             3.0                         4   
2                   DC    Male             2.0                         4   
3                   DC    Male             2.0                         4   
4                   CC    Male             3.0                         3   

  PreferedOrderCat  SatisfactionScore MaritalStatus  NumberOfAddress  \
0           Laptop                  2        Single                9   
1           Mobile                  3        Single                7   
2           Mobile                  3        Single                6   
3           Laptop                  5        Single                8   
4           Mobile                  5        Single                3   

   Complain  OrderAmountHikeFromlastYear  CouponUsed  OrderCount  \
0         1                         11.0         1.0         1.0   
1         1                         15.0         0.0         1.0   
2         1                         14.0         0.0         1.0   
3         0                         23.0         0.0         1.0   
4         0                         11.0         1.0         1.0   

   DaySinceLastOrder  CashbackAmount  
0                5.0          159.93  
1                0.0          120.90  
2                3.0          120.28  
3                3.0          134.07  
4                3.0          129.60

In [ ]:
df.isnull().sum()

CustomerID                     0
Churn                          0
Tenure                         0
PreferredLoginDevice           0
CityTier                       0
WarehouseToHome                0
PreferredPaymentMode           0
Gender                         0
HourSpendOnApp                 0
NumberOfDeviceRegistered       0
PreferedOrderCat               0
SatisfactionScore              0
MaritalStatus                  0
NumberOfAddress                0
Complain                       0
OrderAmountHikeFromlastYear    0
CouponUsed                     0
OrderCount                     0
DaySinceLastOrder              0
CashbackAmount                 0
dtype: int64

In [ ]:
logger.info('Data cleaning ends!')

INFO:root:Data cleaning ends!


# **Save the data to Database**

In [ ]:
from pymongo import MongoClient

In [ ]:
# Connect to Mongodb cluster
client = MongoClient("mongodb+srv://<username>:<password>@mycluster.g3bp8fr.mongodb.net/?retryWrites=true&w=majority")

# List of databases
db_list = client.list_database_names()

In [ ]:
db_list

['CHURN', 'loan_db', 'admin', 'local']

In [ ]:
if 'CHURN' in db_list:
  # Drop database and create new database CHURN
  client.drop_database('CHURN')
  print(client.list_database_names())
  mydb = client.CHURN
else:
  pass

['loan_db', 'admin', 'local']


In [ ]:
# Creating collection/table 'ecomm_churn' in CHURN database
mytb = mydb.ecomm_churn

In [ ]:
# Convert dataframe o dictionary
data = df.to_dict(orient='records') 

# Insert 'data' to ecomm_churn
mytb.insert_many(data)

In [ ]:
# Now check if CHURN added
client.list_database_names()

['CHURN', 'loan_db', 'admin', 'local']

In [ ]:
logger.info('Data saved in database!')

INFO:root:Data saved in database!
